File: Semestrul 5/PPL/Laborator1/src/main/java/ro/ppd2024/Convolution.java
package ro.ppd2024;

public class Convolution {
    public static int convolute(int[][] F, int n, int m, int i, int j, int[][] C, int k) {
        int result = 0, a;

        for (int ii = i - k/2; ii < i + k/2 + 1; ii++) {
            for (int jj = j - k/2; jj < j + k/2 + 1; jj++) {
                if (0 <= ii && ii < n && 0 <= jj && jj < m) {
                    a = F[ii][jj];
                }
                else if (0 < ii && ii < n) {
                    a =  F[ii][jj < 0 ? 0 : m-1];
                }
                else if (0 < jj && jj < m) {
                    a =  F[ii < 0 ? 0 : n-1][jj];
                }
                else {
                    a = F[ii < 0 ? 0 : n-1][jj < 0 ? 0 : m-1];
                }
                result += a;
            }
        }

        return result;
    }
}

File: Semestrul 5/PPD/Laborator1/ConvolutionCpp/util/Convolution.cpp
#include "Convolution.h"
#include <algorithm>

template <typename TMatrix, typename TFilter>
int convolute(TMatrix F, int n, int m, int i, int j, TFilter C, int k) {
    int result = 0;
    int halfK = k / 2;

    for (int ki = -halfK; ki <= halfK; ki++) {
        for (int kj = -halfK; kj <= halfK; kj++) {
            int ni = std::min(std::max(i + ki, 0), n - 1);
            int nj = std::min(std::max(j + kj, 0), m - 1);

            result += F[ni][nj] * C[ki + halfK][kj + halfK];
        }
    }

    return result;
}
